### PR TITLE
[extractor/nebula] Refactor

### PR DESCRIPTION
**Pull request by pukkandan**
* Built on top of https://github.com/yt-dlp/yt-dlp/pull/6334
* Fixes `--flat-playlist` without adding extra requests
* The code had become unnecessarily complex since it was originally written based on Zype. This is a complete refactor
* See https://github.com/yt-dlp/yt-dlp/pull/5979 for previous discussion
* All points in https://github.com/yt-dlp/yt-dlp/pull/5979#issuecomment-1434456017 have been addressed
* Code is only tested without account. Login needs to be tested before merge
	- Verify login doesn't make unnecessary requests
	- Does Re-auth work correctly?
	- ~~Does auth with cookies work?~~ No
	- What is error message for wrong password but correct username?

@hheimbuerger @elyse0

### CODE REVIEW

**Comment by hheimbuerger** [comment]
I believe there is no auth with cookies, because Nebula uses `localStorage`. Unless you have updated information on this? My info is at least two months old, I didn't confirm now.

**Comment by hheimbuerger** [comment]
The `subscriptions` support ([`https://nebula.app/](https://nebula.app/myshows)) is currently broken. I think the `_generate_playlist_entries()` function is still the old one. Were you planning to update that to the new `smuggle_url` architecture?

**Comment by hheimbuerger** [comment]
The unit tests fail for a variety of reasons. I'm in the process of fixing them, but one thing I'm stuck is with the 'download size'. I've seen a parameter for that in someone's PR, but I forgot how to change the expected value here.

```
AssertionError: False is not true : Expected test_Nebula_2_5e779ebdd157bc0001d1c75a.mp4 to be at least 9.77KiB, but it's only 740.00B
```

**Comment by pukkandan** [comment]
> The unit tests fail for a variety of reasons. I'm in the process of fixing them, but one thing I'm stuck is with the 'download size'. I've seen a parameter for that in someone's PR, but I forgot how to change the expected value here.

The test framework doesn't handle hls download well. Just add `'params': {'skip_download': 'm3u8'}` so it only tests metadata

**Comment by pukkandan** [comment]
> The `subscriptions` support ([nebula.app/myshows](https://nebula.app/myshows)) is currently broken. I think the `_generate_playlist_entries()` function is still the old one. Were you planning to update that to the new `smuggle_url` architecture?

I don't remember tbh. I'd have to look at the code again. But sadly, I have been busy irl lately as you can see by my reduced commit history. So I cannot say when I can get back to this. Feel free to look at it yourself if possible.

**Comment by pukkandan** [comment]
I had a quick look, and yes, it's the smuggle_url that is missing. SHould be an easy fix.

**Comment by pukkandan** [comment]
Should be fixed now

File: yt_dlp/extractor/nebula.py [hunk 1]
```diff
@@ -3,139 +3,126 @@
...
-                                            video_id=slug,
-                                            auth_type='bearer',
-                                            note='Fetching video stream info')
-        manifest_url = stream_info['manifest']
-        return self._extract_m3u8_formats_and_subtitles(manifest_url, slug, 'mp4')
+    def _extract_formats(self, slug):
+        stream_info = self._call_api(
+            f'https://content.api.nebula.app/video/{slug}/stream/',
+            slug, note='Fetching video stream info')
+        fmts, subs = self._extract_m3u8_formats_and_subtitles(stream_info['manifest'], slug)
```
Comment (bashonly) [inline yt_dlp/extractor/nebula.py:70]: https://github.com/yt-dlp/yt-dlp/pull/7156#issuecomment-1600301485
```suggestion
        fmts, subs = self._extract_m3u8_formats_and_subtitles(stream_info['manifest'], slug, 'mp4')
```

File: yt_dlp/extractor/nebula.py [hunk 2]
```diff
@@ -3,139 +3,126 @@
...
+                'description': 'description',
+                'timestamp': ('published_at', {parse_iso8601}),
+                'duration': 'duration',
+                'channel_id': 'channel_slug',
+                'uploader_id': 'channel_slug',
+                'channel': 'channel_title',
+                'uploader': 'channel_title',
+                'series': 'channel_title',
+                'creator': 'channel_title',
+                '_old_archive_ids': ('zype_id', {lambda x: [make_archive_id(NebulaIE, x)]}),
```
Comment (bashonly) [inline yt_dlp/extractor/nebula.py:89]: per user feedback, `--download-archive` was particularly useful for subscriptions. Maybe we should make archive IDs for `NebulaIE` and `NebulaSubscriptionsIE`?

Comment (pukkandan) [reply yt_dlp/extractor/nebula.py:89]:
```suggestion
                # Old code was wrongly setting extractor_key from NebulaSubscriptionsIE
                '_old_archive_ids': ('zype_id', {lambda x: [
                    make_archive_id(NebulaIE, x), make_archive_id(NebulaSubscriptionsIE, x)]}),
```

File: yt_dlp/extractor/nebula.py [hunk 3]
```diff
@@ -186,28 +175,58 @@ class NebulaIE(NebulaBaseIE):
...
+        }
 
 
 class NebulaSubscriptionsIE(NebulaBaseIE):
     IE_NAME = 'nebula:subscriptions'
-    _VALID_URL = rf'{_BASE_URL_RE}/myshows'
+    _VALID_URL = rf'{_BASE_URL_RE}/(?P<id>myshows)'
     _TESTS = [
         {
             'url': 'https://nebula.tv/myshows',
```
Comment (seproDev) [inline yt_dlp/extractor/nebula.py:232]: This URL no longer exists. Instead, `https://nebula.tv/library` is used for a general library, with `https://nebula.tv/library/latest-videos` showing the latest videos.

The code currently still works, but the website now uses a new endpoint
```python
'https://content.api.nebula.app/video_episodes/?following=true&include=engagement&ordering=-published_at'
```

File: yt_dlp/extractor/nebula.py [hunk 4]
```diff
@@ -253,30 +275,37 @@ class NebulaChannelIE(NebulaBaseIE):
...
-            episodes = channel['episodes']['results']
+            channel = self._call_api(next_url, collection_id, note=f'Retrieving channel page {page_num}')
 
     def _real_extract(self, url):
         collection_id = self._match_id(url)
-        channel_url = f'https://content.watchnebula.com/video/channels/{collection_id}/'
-        channel = self._call_nebula_api(channel_url, collection_id, auth_type='bearer', note='Retrieving channel')
-        channel_details = channel['details']
+        channel = self._call_api(
+            f'https://content.watchnebula.com/video/channels/{collection_id}/',
```
Comment (seproDev) [inline yt_dlp/extractor/nebula.py:304]: Still works, but site now uses a different endpoint:
```python
f'https://content.api.nebula.app/content/{collection_id}/?include=lessons'
```
Which has a slightly different structure and no longer includes the first set of episode results. Instead, an additional request is made to:
```python
f'https://content.api.nebula.app/video_channels/{channel['channel_id']}/video_episodes/?ordering=-published_at'
```

**Review by seproDev — COMMENTED** [review]
Mostly just comments on how the endpoints used by the website changed. Since the old ones still work, I am not sure if this should be changed.

As for your questions:
> Verify login doesn't make unnecessary requests

Seems like the minimum to me.
>Does Re-auth work correctly?

With the changes from the review it works.
> What is error message for wrong password but correct username?

The response is the same no matter if the account is registered or not.

**Pull request closed without merging**